### ISSUE TITLE
Typescript

### ISSUE THREAD

**Issue by ersimont** [healthiers/mini-dash#7]
As I mentioned in #5, I really like this library's goals. It is almost exactly what I'm looking for. However, I have one important requirement this does not meet: type safety using Typescript. Are you open to including `.d.ts files` in the repository?

**Comment by ersimont** [comment]
I would contribute them at least for the functions I use. However, I don't have a good idea yet how to integrate them into your test setup.

**Comment by bali182** [comment]
Sure, I'm absolutely open to add the d.ts. However since I'm not planning to change the core language of the lib to ts, we'll have to add the definitions manually. I'll look into it, it shouldnt be too difficult considering the size of the lib.

With recent ts versions once we have the d.ts, I think the only thing we have to do is include it in the root of the repo, and you should be able to import from the package properly. As for testing, we definitely should add at some tests to check if the functions accept all the right type of arguments, and don't cause compile errors, functionality is tested ok with the standard js tests.

**Comment by ersimont** [comment]
I'm currently thinking the nicest path forward may be for us to borrow from each-others' libraries. I plan to shamelessly steal tests from yours, and probably some implementations too, maybe over time as I have need for them in my own projects.

You should feel free to steal implementations from mine, too. As for the `.d.ts` files, they are generated automatically for all my functions by the build process, so for any functions where we both have the same signatures - done!  Judging by my `dist/` directory, I'm guessing you would want an `index.d.ts` next to each `index.js`, including the one in the root directory which simply re-exports all the ones in subdirectories.  However, that still does not address testing them, which I should get for "free" as I port tests to typescript (a mistake in the type definitions would cause a compilation error).

Perhaps my next step will be finding a function we have in common, stealing your tests for it, then submitting a PR to get the `.d.ts` file in your repo.

**Comment by ersimont** [comment]
Et voila: https://github.com/simontonsoftware/mini-dash/commit/c1945b10eaa4222e31a1973c3e052b42b02448e8

I _think_ this is how it should be set up. Let me know if you think this is worth continuing, and I can add more like it.

**Comment by bali182** [comment]
Types needs some refinement, but looks good! I assume there would be an `index.d.ts` that simply imports and re-exports everything (I know definitions don't really import/export stuff, but that's how it would look like in the code)?

Also, don't be ashamed "steal" whatever you need 😄 but as we discussed some methods are nowhere near feature equivalent with lodash, so changes might be required.

**Comment by ersimont** [comment]
Yes, there would be an `index.d.ts` that imports & re-exports everything ... and there already is!  When you looked at this it simply had the line `export * from './clone';`.  I just committed another one, so it should have 2 lines now.
